src/teachy_tv.c

```diff
@@ -0,0 +1,1668 @@
+        ldr r2, _0815BFEC @ =0xfffffd80\n\
+        adds r1, r2\n\
+    _0815BF96:\n\
+        lsls r1, 4\n\
+        ldr r0, [r0, 0xC]\n\
+        adds r5, r0, r1\n\
+        ldrh r1, [r5]\n\
+        adds r0, r6, 0\n\
+        bl TeachyTvComputePalIndexArrayEntryByMetaTile\n\
+        lsls r0, 12\n\
+        lsls r4, 2\n\
+        adds r0, r4\n\
+        strh r0, [r7]\n\
+        ldrh r1, [r5, 0x2]\n\
+        adds r0, r6, 0\n\
+        bl TeachyTvComputePalIndexArrayEntryByMetaTile\n\
+        lsls r0, 12\n\
+        adds r0, r4\n\
+        adds r0, 0x1\n\
+        strh r0, [r7, 0x2]\n\
+        ldrh r1, [r5, 0x4]\n\
+        adds r0, r6, 0\n\
+        bl TeachyTvComputePalIndexArrayEntryByMetaTile\n\
+        adds r1, r7, 0\n\
+        adds r1, 0x40\n\
+        lsls r0, 12\n\
+        adds r0, r4\n\
+        adds r0, 0x2\n\
+        strh r0, [r1]\n\
+        ldrh r1, [r5, 0x6]\n\
+        adds r0, r6, 0\n\
+        bl TeachyTvComputePalIndexArrayEntryByMetaTile\n\
+        adds r1, r7, 0\n\
+        adds r1, 0x42\n\
+        lsls r0, 12\n\
+        adds r0, r4\n\
+        adds r0, 0x3\n\
+        strh r0, [r1]\n\
+        pop {r4-r7}\n\
+        pop {r0}\n\
+        bx r0\n\
+        .align 2, 0\n\
+    _0815BFEC: .4byte 0xfffffd80\n\
+        ");
+}
+#endif
+
+void TeachyTvComputeMapTilesFromTilesetAndMetaTiles(u16 *metaTilesArray, u8 *blockBuf, u8 *tileset)
+{
+    TeachyTvComputeSingleMapTileBlockFromTilesetAndMetaTiles(blockBuf, &tileset[0x20 * (*metaTilesArray & 0x3FF)], (*metaTilesArray >> 10) & 3);
+    TeachyTvComputeSingleMapTileBlockFromTilesetAndMetaTiles(blockBuf, &tileset[0x20 * (metaTilesArray[4] & 0x3FF)], (metaTilesArray[4] >> 10) & 3);
+    TeachyTvComputeSingleMapTileBlockFromTilesetAndMetaTiles(blockBuf + 0x20, &tileset[0x20 * (metaTilesArray[1] & 0x3FF)], (metaTilesArray[1] >> 10) & 3);
+    TeachyTvComputeSingleMapTileBlockFromTilesetAndMetaTiles(blockBuf + 0x20, &tileset[0x20 * (metaTilesArray[5] & 0x3FF)], (metaTilesArray[5] >> 10) & 3);
+    TeachyTvComputeSingleMapTileBlockFromTilesetAndMetaTiles(blockBuf + 0x40, &tileset[0x20 * (metaTilesArray[2] & 0x3FF)], (metaTilesArray[2] >> 10) & 3);
+    TeachyTvComputeSingleMapTileBlockFromTilesetAndMetaTiles(blockBuf + 0x40, &tileset[0x20 * (metaTilesArray[6] & 0x3FF)], (metaTilesArray[6] >> 10) & 3);
+    blockBuf += 0x60;
+    TeachyTvComputeSingleMapTileBlockFromTilesetAndMetaTiles(blockBuf, &tileset[0x20 * (metaTilesArray[3] & 0x3FF)], (metaTilesArray[3] >> 10) & 3);
+    TeachyTvComputeSingleMapTileBlockFromTilesetAndMetaTiles(blockBuf, &tileset[0x20 * (metaTilesArray[7] & 0x3FF)], (metaTilesArray[7] >> 10) & 3);
+}
+
+#ifdef NONMATCHING
+void TeachyTvComputeSingleMapTileBlockFromTilesetAndMetaTiles(u8 *blockBuf, u8 *tileset, u8 metaTile)
+{
+    u8 *buffer;
+    u32 i;
+    u32 j;
+    vu32 src;
+
+    buffer = (u8 *)AllocZeroed(0x20);
+    src = ((u32)AllocZeroed(0x20));
+    CpuFastSet(tileset, buffer, 8);
+    if ( metaTile & 1 )
+    {
+        i = 0;
```

Comment:
this is almost guaranteed to be a `for` loop with `u8 i, j;`

Reply:
it got rewritten but still failed